test/common/init/target_impl_test.cc

```diff
@@ -60,6 +60,104 @@ TEST(InitTargetImplTest, ReadyWhenWatcherUnavailable) {
   EXPECT_FALSE(target.ready());
 }
 
+// SharedTarget acts as TargetImpl if single watcher is provided.
+TEST(InitSharedTargetTestImpl, InitializeSingleWatcher) {
+  InSequence s;
+
+  ExpectableSharedTargetImpl target;
+  ExpectableWatcherImpl watcher;
+
+  // initializing the target through its handle should invoke initialize()...
+  target.expectInitialize();
+  EXPECT_TRUE(target.createHandle("test")->initialize(watcher));
+
+  // calling ready() on the target should invoke the saved watcher handle...
+  watcher.expectReady();
+  EXPECT_TRUE(target.ready());
+
+  // calling ready() a second time should have no effect.
+  watcher.expectReady().Times(0);
+  target.ready();
+}
+
+// Initializing TargetHandle return false if uninitialized SharedTarget is destroyed.
+TEST(InitSharedTargetTestImpl, InitializeWhenUnavailable) {
+  InSequence s;
+  ExpectableWatcherImpl watcher;
+  TargetHandlePtr handle;
+  {
+    ExpectableSharedTargetImpl target;
+
+    // initializing the target after it's been destroyed should do nothing.
+    handle = target.createHandle("test");
+    target.expectInitialize().Times(0);
+  }
+  EXPECT_FALSE(handle->initialize(watcher));
+}
+
+// Initializing TargetHandle return false if initialized SharedTarget is destroyed.
+TEST(InitSharedTargetTestImpl, ReInitializeWhenUnavailable) {
+  InSequence s;
+  ExpectableWatcherImpl w;
+  TargetHandlePtr handle;
+  {
+    ExpectableSharedTargetImpl target;
+
+    target.expectInitialize();
+    TargetHandlePtr handle1 = target.createHandle("m1");
+    ExpectableWatcherImpl w1;
+    EXPECT_TRUE(handle1->initialize(w1));
+
+    // initializing the target after it's been destroyed should do nothing.
+    handle = target.createHandle("m2");
+    target.expectInitialize().Times(0);
+    // target destroyed
+  }
+  EXPECT_FALSE(handle->initialize(w));
+}
+
+// SharedTarget notifies multiple watchers.
```

Comment:
So these two new tests capture these two sequences, if I'm reading correctly:

```
initialize(w1)
initialize(w2)
ready
```

and

```
initialize(w1)
ready
initialize(w2)
```

I guess you could also write one for 

```
ready
initialize(w1)
initialize(w2)
```

But that's probably not very interesting... I can't think of any more interesting cases, so I think you're good. Can you think of anything?

Reply:
It's not hard to support a ready() prior to initialize any watcher.
The confusing part what are the expected consequences.
MUST: notify the following watchers 
Question: Should the initialization function defined in SharedTarget should be called?

I am going to leave the questionable behavior as undefined and set no expectation. 
Let me know what you think